### ISSUE TITLE
Catch OSM that is excessively large or crosses the 180 degree meridian

### ISSUE THREAD

**Issue by abyrd** [conveyal/r5#738]
This problem was encountered on an Auckland, New Zealand project which had been set up using OSM data for the entire country. When a single point analysis was requested, the worker eventually failed with an out of memory error. Examining logs revealed the following:

```
LinkedPointSet - Linked 9680000 of 1486431408 PointSet points to streets.
WebMercatorGridPointSet - Creating WebMercatorGridPointSet with WGS84 extents Env[-177.9799847 : 178.5482951, -52.5908055 : -29.2418316]
```

Although the region and analysis bounds were set to a relatively small area around Auckland, the worker was dutifully linking a grid of about 1.5 billion cells, spanning the whole height of New Zealand and almost the full width of the Earth. 

There are two reasons: First, when building a network, the worker will build a base linkage covering the full extents of the OSM network, not considering the actually requested bounds of analysis. This is intentional, as it allows changing the bounds of analysis later, but does allow for extreme mismatches between the two bounding boxes when uncropped OSM inputs are used. In this case, the presence of outlying islands makes the dataset around 10 times larger in the north-south direction than it would need to be. We should try to detect such extreme mismatches and advise the user to crop and filter the input data.

Second, the dataset for all of New Zealand also includes some outlying islands on the other side of the 180 degree line of longitude. When the grid's bounding box is expanded to include the whole supplied OSM data set, it does not consider the possibility of creating a more compact bounding box by spanning the 180 degree meridian. This is not a problem in itself, because it's much simpler to work with envelopes that are defined not to cross 180 degrees, and we are unlikely to encounter analysis regions that actually cross that line. But we do need to at least detect data sets violating these rules and provide appropriate guidance to the user when they are spotted.

**Comment by abyrd** [comment]
Currently our main line of defense against these problematic data sets is the static method `Grid#checkWgsEnvelopeSize`. This is called when creating opportunity data grids from shapefile or csv, in `FreeFormPointSet#fromCsv`, and in the DataSourceIngesters for GeoJson, GeoPackage, and Shapefile. There are two shortcomings here: first, it only checks the surface area of the envelope, so doesn't throw an error specifically mentioning the 180 degree meridian. We should specifically check whether the x span is made smaller by ofsetting the minimum x value by 180 degrees. Second, we're only checking on data source upload and opportunity data / freeform pointset creation.  We should also check on OSM and GTFS upload and linking. OSM and GTFS will eventually be datasource uploads but better to get this check in sooner than later.

**Comment by abyrd** [comment]
This should be fixed by 8d2a6e0, to be included in release v6.5.
